### PR TITLE
fix(strava): skip 404/403 stream fetch errors instead of aborting sync

### LINKED ISSUE #65: strava-data-puller: stream fetch crashes on 404 (activity 15947122602)
## Bug: Unhandled 404 in `fetch_activity_streams`

During daily ingestion on 2026-06-19, `strava_pull.py --include-streams --backfill-details` crashed with a 404 HTTP error while backfilling streams for activity `15947122602`.

### Error

```
requests.exceptions.HTTPError: 404 Client Error: Not Found for url:
https://www.strava.com/api/v3/activities/15947122602/streams?keys=time%2C...
```

Stack trace via `fetch_activity_streams` → `request_json` → `response.raise_for_status()`.

### Root Cause

`request_json()` calls `raise_for_status()` unconditionally. Some activities (likely deleted, private, or not owned by the token holder) return 404 for their streams endpoint. This aborts the entire sync rather than skipping the problematic activity.

### Impact

- Stream backfill for all subsequent activities in the queue is skipped
- Parquet export is not reached, so raw/curated data is not updated during the run that hits the 404

### Fix

Catch `HTTPError` with status 404 (and possibly 403) in `fetch_activity_streams`, log a warning, and continue. Optionally write a sentinel file so the activity isn't retried every run.

```python
try:
    streams = request_json(...)
except requests.exceptions.HTTPError as exc:
    if exc.response.status_code == 404:
        print(f"Warning: streams not available for activity {activity_id} (404), skipping.")
        return
    raise
```

### Workaround

Run without `--include-streams` for daily syncs (basic activity list still works fine).

### PR DESCRIPTION
**Pull request by mohit**
## Summary
Catch HTTPError 404 (and 403) in fetch_activity_streams and skip the problematic activity with a warning instead of aborting the entire sync run.

## Verification
- Reviewed fetch_activity_streams and request_json call path
- Minimal exception handler added; non-404/403 errors still propagate
- Existing tests pass (or note if no tests present)

## Fixes
Fixes mohit/tools#65

### CODE REVIEW

**Review by mohit — COMMENTED** [review]
Addressed in 5329082: on 403/404, `fetch_activity_streams` now calls `cached.unlink()` on `streams/<activity_id>.json` before returning `None`, so stale cached data is never re-indexed by `build_activity_streams_ndjson` / `main()`. Also added `test_fetch_activity_streams_deletes_cached_file_on_403_404` to cover both status codes. All 13 tests pass.

File: strava-data-puller/strava_pull.py [hunk 1]
```diff
@@ -425,17 +425,28 @@ def fetch_activity_streams(
...
+        )
+    except requests.exceptions.HTTPError as exc:
+        status = exc.response.status_code if exc.response is not None else None
+        if status in (403, 404):
+            print(
+                f"Warning: streams not available for activity {activity_id} "
+                f"({status}), skipping.",
+                file=sys.stderr,
+            )
+            return None
```
[the remark this answers is not in the document]
Comment (mohit) [reply strava-data-puller/strava_pull.py]: Fixed in 5329082: on 403/404, the handler now calls `cached.unlink()` on `streams/<activity_id>.json` before returning `None`, and logs a warning to stderr. This ensures `build_activity_streams_ndjson()` in `main()` won't pick up stale stream data for activities the API has declared unavailable.

**Comment by mohit** [comment]
🐻 **Cron review pass** — 2026-06-19 22:21Z

**CI:** 7/8 pass ✅ — lint ❌ (1 failure)
- All `test (*)` jobs: **pass**
- `lint`: **fail** — `SIM103` at `strava_pull.py:548`: *Return the condition directly*

⚠️ **Lint failure is pre-existing** — this SIM103 violation exists on `main` at line 530 (shifted to 548 by the +18 lines this PR adds to `fetch_activity_streams`). Not introduced by this PR.

**Diff review:**

✅ **404/403 skip logic** — correct. `requests.exceptions.HTTPError` is caught, status code checked; only 403/404 are silenced, everything else re-raises. Normal sync is unaffected.

✅ **Cached file cleanup** — correct and complete. `streams/<activity_id>.json` is unlinked before returning `None`. Stale re-index risk eliminated. Addresses inline comment (Fixed in 5329082 ✓).

✅ **Return type widened to `dict | None`** — callers at lines 894 and 911 discard the return value entirely, so `None` on skip is safely ignored; sync proceeds to the next activity.

✅ **Tests** — cover both 403 and 404 paths, verify cached file deletion, assert `None` return. Solid coverage.

**Merge readiness:** Functionally ✅ ready. The lint block is pre-existing noise unrelated to this fix — suggest a follow-up commit `fix(lint): SIM103 in activity_in_range` or squash it in before merge to clear the UNSTABLE state. Up to you whether to block on it.

**Comment by mohit** [comment]
🐻 **Cron review pass** — 2026-06-20 07:21 PDT

**CI:** ✅ All 8 checks green (lint, test (strava-data-puller), and 6 others)

**Commits since last review pass (`7382a3f`):**
- `5329082` fix(strava): remove cached stream file on 404/403 skip to prevent stale re-index
- `d95774d` fix(lint): simplify activity_in_range return (SIM103)

**Diff assessment — `strava-data-puller/strava_pull.py`:**

1. **`fetch_activity_streams` (404/403 handling):** Return type widened to `dict[str, Any] | None`. On 403/404, logs a warning to stderr, removes any stale `streams/<id>.json` cache file, and returns `None`. All callers (lines 892, 909) ignore the return value, so the `None` path is safe. On other HTTP errors the exception is re-raised — correct behaviour. Stale-cache cleanup addresses the P2 bot comment (resolved in `5329082`).

2. **`activity_in_range` / `activity_in_scope` (SIM103):** `return not (before is not None and timestamp > before)` is a clean mechanical simplification — logically equivalent, no behaviour change.

**Inline comments:** The one P2 bot comment (clear cached streams on 403/404 skip) is resolved by `5329082`. No outstanding unresolved comments.

**Issues found:** None. Code is correct and clean.

✅ **All clear — merge-ready.**

**Comment by mohit** [comment]
Verified commit 5329082: on 403/404, `fetch_activity_streams` now calls `cached.unlink()` on `streams/<activity_id>.json` before returning `None`, and logs a warning to stderr — fully addressing the P2 thread. New test `test_fetch_activity_streams_deletes_cached_file_on_403_404` covers both status codes. CI is green (8/8). PR is ready to merge.

**Review by mohit — COMMENTED** [review]
## Code Review — fix/issue-65 (d95774d6)

**TL;DR: Ready to merge.** The fix is correct and complete. A few minor observations below.

---

### Fix correctness ✅

`fetch_activity_streams` now wraps `request_json` in a `try/except requests.exceptions.HTTPError`. On 403/404 it logs to stderr, unlinks any stale cache file, and returns `None`. All other HTTP errors are re-raised, preserving prior behavior. The `exc.response is not None` guard before reading `.status_code` is important and is handled correctly (falls through to re-raise when response is missing).

### Cache-unlink safety ✅

`cached.exists()` is checked before `cached.unlink()` — no risk of `FileNotFoundError` when no cache is present. Both log lines go to stderr as expected.

### Caller compatibility ✅

Both call sites (new-activity path ~L892, backfill path ~L909) ignore the return value, so the `None` return is silently swallowed and the sync loop continues to the next activity. `details_or_streams_updated` is still set correctly in both paths.

### Test coverage ✅ (with minor gaps)

New test `test_fetch_activity_streams_deletes_cached_file_on_403_404` covers both status codes, verifies `None` return, and asserts the cached file is deleted. Two small gaps worth noting (not blocking):
- No test for the "no cached file present" path (code is safe, just untested).
- No test asserting a non-403/404 `HTTPError` (e.g., 500) is re-raised rather than swallowed.

### Remaining edge cases (future work, not blocking)

- **401 Unauthorized** — would re-raise and crash the sync. Reasonable for now; a token-refresh retry could be a follow-up.
- **5xx / network timeouts** — re-raised per existing behavior, fine as-is.
- The `activity_in_scope` boolean simplification is a no-op refactor; no concerns.

**Verdict: approve and merge.** 🟢

**Review by mohit — COMMENTED** [review]
🐻 **Cron review pass (Bear Claw)**

CI: ✅ all 8 green. Code verified: `fetch_activity_streams` catches HTTPError, calls `cached.unlink()` on 403/404 with stderr warning, re-raises on other errors. Local test suite: 13/13 pass. P2 cached-streams issue fully addressed. ✅ Ready to merge.

**Review by mohit — COMMENTED** [review]
🐻 **Cron review pass (Bear Claw)** — CI: ✅ all 8 green. Code verified: cached.unlink() on 403/404 ✅, stderr warning ✅, non-4xx re-raises ✅. Local tests: 13/13 pass. P2 addressed. Ready to merge.

**Comment by mohit** [comment]
🐻 **Cron review pass (Bear Claw)**

CI: ✅ all 8 green. Code verified: `fetch_activity_streams` catches HTTPError, calls `cached.unlink()` on 403/404 with stderr warning, re-raises on other errors. Local tests: 13/13 pass. P2 cached-streams issue fully addressed. ✅ Ready to merge.

**Comment by mohit** [comment]
🐻 **Cron review pass (Bear Claw)** — CI: ✅ all 8 green — No new comments or changes since last pass; 403/404 skip + cached.unlink() fix confirmed, all prior review threads resolved. Ready to merge.
